src/utils/io/progress_log.rs

```diff
@@ -0,0 +1,138 @@
+// Copyright 2019-2023 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use humantime::format_duration;
+use std::time::{Duration, Instant};
+
+use parking_lot::Mutex;
+use pin_project_lite::pin_project;
+use std::io;
+use std::pin::Pin;
+use std::sync::Arc;
+use std::task::{Context, Poll};
+use tokio::io::ReadBuf;
+
+use log::info;
+
+const UPDATE_FREQUENCY: Duration = Duration::from_millis(5000);
+
+pin_project! {
+    /// Wraps an iterator to display its progress.
+    pub struct WithProgressStream<S> {
+        #[pin]
+        stream: S,
+        progress: WithProgress,
+    }
+}
```

Comment:
I can see why we need a separate struct - it allows us to have a `ProgressLog` a `WithProgressStream`, with the implementation detail of `WithProgress` common.

But I think we should have only one struct, if not in this PR, then in a clean up for the issues below.


Here's the API I think we should be going for
```rust
pub struct WithProgress<Inner> {
    inner: Inner,
    known_total: Option<u64>,
    seen: u64,
    last_logged: Instant,
    message: String,
}

impl<Inner> WithProgress<Inner> {
    pub fn new(inner: Inner, message: impl Display);
    pub fn new_with_known_total(inner: Inner, message: impl Display, known_total: u64);
    fn update_and_log_if_required(&mut self, newly_seen: u64);
}

// as required by user code
impl<Inner: Iterator> Iterator for WithProgress<Inner> {}
impl<Inner: AsyncRead> AsyncRead for WithProgress<Inner> {}
impl<Inner: AsyncWrite> AsyncWrite for WithProgress<Inner> {}
impl<Inner: Stream> Stream for WithProgress<Inner> {}

```

Reply:
I'm curious why won't we do something like: 
```
impl<Inner> WithProgress<Inner> {
    pub fn new(inner: Inner) -> Self;
    pub fn with_message(&self, message) -> Self;
    pub fn with_total(&self, total);
    fn update_and_log_if_required(&mut self, newly_seen: u64);
}
```

It's a much nicer api than verbose constructors and much more flexible too.

Reply:
That also works!

Reply:
Yeah, message should probably be required, unless we want this to be flexible for some reason. No strong opinion there, would just prefer to have 1 constructor instead of several. And for optional params use this pattern. I'm still not sure if there's ever a time when we don't know the `total_size` at init.